internal/pkg/gateway/api_test.go

```diff
@@ -883,11 +899,41 @@ func TestEndorse(t *testing.T) {
 			},
 			expectedEndorsers: []string{"localhost:7051", "peer2:9051"},
 		},
+		{
+			name: "context timeout during first endorsement",
+			plan: endorsementPlan{
+				"g1": {{endorser: localhostMock, height: 3}}, // msp1
+				"g2": {{endorser: peer4Mock, height: 5}},     // msp3
+			},
+			timeout:   time.Nanosecond,
+			errString: "rpc error: code = DeadlineExceeded desc = endorsement timeout expired",
+		},
+		{
+			name: "context timeout collecting endorsements",
+			plan: endorsementPlan{
+				"g1": {{endorser: localhostMock, height: 3}}, // msp1
+				"g2": {{endorser: peer4Mock, height: 5}},     // msp3
+			},
+			timeout: 100 * time.Millisecond,
+			postSetup: func(t *testing.T, def *preparedTest) {
+				peer4Mock.client.(*mocks.EndorserClient).ProcessProposalStub = func(ctx context.Context, proposal *peer.SignedProposal, option ...grpc.CallOption) (*peer.ProposalResponse, error) {
+					time.Sleep(200 * time.Millisecond)
+					return createProposalResponse(t, peer3Mock.address, "mock_response", 200, ""), nil
+				}
+			},
+			errString: "rpc error: code = DeadlineExceeded desc = endorsement timeout expired",
+		},
```

Comment:
These timing based racing tests are prone to add flakes when they run on different hardware (like the first test here) or they add unnecessary delays (like the second test). Can you do change them to somewhat on the following lines...

- Instead of `timeout` store the context in the testdef
- In the mock endorser, call `cancel()` on the `context` and then return after a small wait

Also, can you change the error message slightly so the message from the first endorsement is different  from when you collect the endorsements in parallel - so as to make sure that the error is coming from the intended place.

Reply:
Made both changes. Thanks!